examples/step-19/step-19.cc

```diff
@@ -0,0 +1,728 @@
+                                       QGauss<dim - 1>(fe.degree + 1),
+                                       {},
+                                       solution,
+                                       estimated_error_per_cell);
+
+    GridRefinement::refine_and_coarsen_fixed_number(triangulation,
+                                                    estimated_error_per_cell,
+                                                    0.1,
+                                                    0.03);
+
+    triangulation.execute_coarsening_and_refinement();
+  }
+
+
+  // @sect4{CathodeRaySimulator::create_particles}
+
+  template <int dim>
+  void CathodeRaySimulator<dim>::create_particles()
+  {
+    FEFaceValues<dim>           fe_face_values(fe,
+                                     QMidpoint<dim - 1>(),
+                                     update_quadrature_points |
+                                       update_gradients |
+                                       update_normal_vectors);
+    std::vector<Tensor<1, dim>> solution_gradients(
+      fe_face_values.n_quadrature_points);
+    const FEValuesExtractors::Scalar electric_potential(0);
+
+    types::particle_index n_current_particles =
+      particle_handler.n_global_particles();
+
+    for (const auto &cell : dof_handler.active_cell_iterators())
+      for (const auto &face : cell->face_iterators())
+        if (face->at_boundary() && face->boundary_id() == BoundaryIds::cathode)
+          {
+            // compute the electric field strength at each quadrature point
+            fe_face_values.reinit(cell, face);
+            fe_face_values[electric_potential].get_function_gradients(
+              solution, solution_gradients);
+            for (const unsigned int q_point :
+                 fe_face_values.quadrature_point_indices())
+              {
+                const Tensor<1, dim> E = solution_gradients[q_point];
+
+                // electrons can only escape the cathode if the electric field
+                // strength exceeds a threshold and,
+                // crucially, if the electric field points *into* the domain
+                if ((E * fe_face_values.normal_vector(q_point) < 0) &&
+                    (E.norm() > Constants::E_threshold))
+                  {
+                    const Point<dim> location =
+                      fe_face_values.quadrature_point(q_point);
+                    // create a particle at 'location' and insert it into the
+                    // ParticleHandler
+
+                    Particles::Particle<dim> new_particle;
+                    new_particle.set_location(location);
+                    new_particle.set_reference_location(
+                      mapping.transform_real_to_unit_cell(cell, location));
+                    new_particle.set_id(n_current_particles);
+                    particle_handler.insert_particle(new_particle, cell);
+
+                    ++n_current_particles;
```

Comment:
This variable is currently unused. Is it useful?

Reply:
I use this to assign ids three lines above. But the logic is wrong once particles get lost. Will fix.